src/dialogs/restart/dialog-restart.ts

```diff
@@ -232,84 +264,87 @@ class DialogRestart extends LitElement {
+    switch (this._backupState) {
+      case "create_backup":
+        return this.hass.localize("ui.dialogs.restart.backup_in_progress");
+      case "receive_backup":
+        return this.hass.localize("ui.dialogs.restart.upload_in_progress");
+      case "restore_backup":
+        return this.hass.localize("ui.dialogs.restart.restore_in_progress");
+      default:
+        return "";
     }
+  }
 
-    this.closeDialog();
+  private _restartAction =
+    (
+      domain: ServiceCallRequest["domain"],
+      service: ServiceCallRequest["service"],
+      errorTitle: string,
 
-    try {
-      await this.hass.callService("homeassistant", "restart");
-    } catch (err: any) {
-      showAlertDialog(this, {
-        title: this.hass.localize("ui.dialogs.restart.restart.failed"),
-        text: err.message,
-      });
-    }
-  }
+      serviceData?: ServiceCallRequest["serviceData"]
+    ) =>
+    async () => {
+      try {
+        await this.hass.callService(domain, service, serviceData);
+      } catch (err: any) {
+        showAlertDialog(this, {
+          title: errorTitle,
+          text: err.message,
+        });
+      }
+    };
 
-  private async _showRestartSafeModeDialog() {
-    const confirmed = await showConfirmationDialog(this, {
-      title: this.hass.localize(
-        "ui.dialogs.restart.restart-safe-mode.confirm_title"
-      ),
-      text: this.hass.localize(
-        "ui.dialogs.restart.restart-safe-mode.confirm_description"
-      ),
-      confirmText: this.hass.localize(
-        "ui.dialogs.restart.restart-safe-mode.confirm_action"
-      ),
-      destructive: true,
-    });
+  private _hostAction =
+    (toastMessage: string, action: "reboot" | "shutdown") => async () => {
+      showToast(this, {
+        message: toastMessage,
+        duration: -1,
+      });
 
-    if (!confirmed) {
-      return;
-    }
+      try {
+        if (action === "reboot") {
+          await rebootHost(this.hass);
+        } else {
+          await shutdownHost(this.hass);
+        }
+      } catch (err: any) {
+        // Ignore connection errors, these are all expected
+        if (this.hass.connection.connected && !ignoreSupervisorError(err)) {
+          showAlertDialog(this, {
+            title: this.hass.localize(`ui.dialogs.restart.${action}.failed`),
+            text: extractApiErrorMessage(err),
+          });
+        }
+      }
+    };
 
-    this.closeDialog();
+  private async _handleAction(ev) {
+    const action = ev.currentTarget.action as
+      | "restart"
+      | "reboot"
+      | "shutdown"
+      | "restart-safe-mode";
 
-    try {
-      await this.hass.callService("homeassistant", "restart", {
-        safe_mode: true,
-      });
-    } catch (err: any) {
-      showAlertDialog(this, {
-        title: this.hass.localize(
-          "ui.dialogs.restart.restart-safe-mode.failed"
-        ),
-        text: err.message,
-      });
-    }
-  }
+    const backupProgressMessage = this._getBackupProgressMessage();
```

Comment:
Why dont we call `_loadBackupState` here, as it is when we need it and use it, and it could be that a backup was started (by another user, automatic) between opening the dialog and showing the confirm dialog.

Reply:
Yes this makes sense, I did load it in the beginning because with my supervisor remote it took some time to load. But I changed it to show a loading indicator on pressing the action